### ISSUE TITLE
White-/Blacklist for sitemap parameters

### ISSUE THREAD

**Issue by KreativeKrise** [mblaschke/TYPO3-metaseo#254]
Would it be possible to add a constant where we can set a white- or blacklist for GET parameters? Often there are URLs with tracking parameters like "utm_source".

It would be nice to exclude URLs with such parameters from the sitemap generation.

Or is that already possible? I didn't find any information..

Thank you in advance!

**Comment by thomaszbz** [comment]
Which version of metaseo, TYPO3 and PHP are you using? Do you have realurl enabled and which version do you use?

I get the idea: These parameters can lead to duplicate content (multiple URLs with same content). For instance, you can also exclude these parameters at e.g. google search console and e.g. piwik analytics (while the latter does not use metaseo's sitemap).

How exactly do you create these get parameters? Using an extension? Does TYPO3 _know_ something about these parameters or do you just add them outside of the scope of TYPO3?

I need to reproduce this issue first, to see if it really happens and if it creates _new_ entries with 2.0.0. Did you try this out already?

What happens with your canonical tag when you use such parameters?

**Comment by KreativeKrise** [comment]
Metaseo 1.0.8
TYPO3 6.2.25
PHP 5.4.41

Realurl is enabled.

These parameters are created by an external newsletter service. It adds the tracking parameters for Google Analytics to the links inside the newsletter.

I will try version 2.0.0 on Monday if it maybe fixes the problem!
The canonical tag I have to check also on Monday because I don't use the functionality of metaseo.
So I will response on Monday!

Have a great weekend :)

**Comment by thomaszbz** [comment]
I'm not sure if 2.0.0 fixes the problem, but we fixed many many bugs since 1.0.8. 1.0.x is not supported any more (We still support 6.2 with 2.0.0).

But there's one regression which is not yet fixed in 2.0.0: #233 in case you have a full-blown hoster's environment, together with TYPO3 6.2. A fix is available in metaseo's develop branch, however.

**Comment by KreativeKrise** [comment]
I tried 2.0.0 and URLs with "wrong" GET-parameter generate an entry inside the sitemap.xml.
An option similar to `metaTags.canonicalUrl.strict = 1` for the sitemap.xml would be great. Even better would be a whitelist for GET-parameters :)

**Comment by thomaszbz** [comment]
What happens if you use a parmeter which is new to TYPO3 (not configured somewhere)?

E.g. `http://www.example.com/some-page.html?thisisnew=1`

**Comment by KreativeKrise** [comment]
If I add a parameter like `?thisisnew=1` then it generates a new sitemap entry.
The canonical tag is fine, as long as I use `metaTags.canonicalUrl.strict = 1`.

**Comment by thomaszbz** [comment]
I've been able to reproduce this issue in the meantime, at least with parameters which are new to TYPO3.
![auswahl_040](https://cloud.githubusercontent.com/assets/11200727/18476628/190840bc-79cb-11e6-9f85-bcd27f03a88a.png)

**Comment by thomaszbz** [comment]
I retried this with 6.2.29, meaning that #268 was applied (including configuration change). With the security-patched versions of TYPO3 I don't get new entries when I request pages with parameters like `/?pk_campaign=abc`.

If someone else still is affected by this issue, please

- wipe out unwanted entries from your sitemap e.g. via metaseo's sitemap editing functionality (delete/blacklist whereas delete should now be appropriate)
- make sure that you apply #268, including the change of configuration (you need to do this manually!).

Basically, that means we do well for

- sitemap via checking against TYPO3's cache
- canonical URL in strict mode, according to the feedback provided by @KreativeKrise.

For the moment, I think this bug is fixed via the core #268, closing this respectively. If someone still is affected with #268 applied, please open a new issue and reference back to this one.

**Comment by sebastianschrama** [comment]
Please reopen this issue, as this problem already exists for me.

TYPO3 7.6.15
metaseo 2.0.3
realurl 2.1.5
php 5.6.27

`cHashIncludePageId` is set to `1`

I already have parameters like "gclid", which comes from google adwords in the sitemap.xml. I also tested with other parameters (?parameter=test). Same problem.

What's wrong? Did I miss setting up some options? Or is there still a bug?

**Comment by thomaszbz** [comment]
@sebastianschrama Thanks for reporting back.

My test was against TYPO3 6.2.29. Considered that you are still faced with this issue, closing this issue should indeed go along with a test against 7.6.15+.

It's unlikely I can do something before March 2017 (limited time). Would be great if someone could track this issue further down in the meantime.

To answer your questions:

* Please make sure that you cleared all caches (including opcode cache via install tool)
* Make sure that development mode is disabled
* Clean your sitemap via Metaseo's backend module (delete entries)
* Clear caches again
* Log out from backend
* NOW run your test again.

It's a bug if parameters unknown to TYPO3 make it to the sitemap again. If these entries also blow up TYPO3's caches (unlimited), it could even be a security issue in the core.

**Comment by sebastianschrama** [comment]
Thanks for the quick response. I followed your instructions, but unfortunately the sitemap still adds unknown parameters in urls.

**Comment by thomaszbz** [comment]
Maybe @benf can help out? @benf I already took into account that your patch set could eventually be useful for this issue.

**Comment by thomaszbz** [comment]
@sebastianschrama Other user says this is fixed against 7.6: https://github.com/mblaschke/TYPO3-metaseo/issues/306#issuecomment-272071718

@sebastianschrama Please make sure that `cHashIncludePageId` is indeed active.

**Comment by sebastianschrama** [comment]
The `cHashIncludePageId` is definitely set to `1` in my case.

**Comment by thomaszbz** [comment]
I know. Deleted my previous message long before you answered ;-)

**Comment by sebastianschrama** [comment]
Any news here?

**Comment by thomaszbz** [comment]
Would be great if someone could provide a patch or track this down at least (=> "help wanted" tag). I still don't have a lot of time currently.

**Comment by danpii** [comment]
Moderating this issue, I moved this comment to #505.

@thomaszbz

**Comment by thomaszbz** [comment]
@DanPii That sounds cooluri related. I moved it to a separate ticket #505.

**Comment by danpii** [comment]
@thomaszbz I think only the URLs containing ADMCMD_cooluri=1 are cooluri relevant.
- ?fb_locale=nl_NL --> looks like like a facebook url parameter to me
- ?ad=semD... --> looks like an url parameter from some ad site (Google Ads?, Adwords?)

I am not using typo3 extensions for FB or Google Ads, but facebook javascript on some pages, so Typo3 will not know about these parameters.

Also, I just saw in the documentation that there is a Typoscript Node called:
plugin.metaseo.sitemap.index.blacklist
Could I use that with regular expressions to put the relevant links on a blacklist?

Thanks
Daniel

**Comment by thomaszbz** [comment]
@DanPii Well, for the case it's not CoolUri related, then it just duplicates this issue.

You can blacklist these entries in the sitemap view of metaseo at least.

**Comment by danpii** [comment]
Blacklisting with the sitemap view of metaseo only works after they have been already added. Unfortunately there were a lot.
Is it possible via Typoscript and this Node "plugin.metaseo.sitemap.index.blacklist" to avoid them being added in the first place?

**Comment by thomaszbz** [comment]
@DanPii Currently, I'm not aware of a workaround for this issue. If you find a workaround for this issue, please post it here. Maybe we can use it for the fix as well.

**Comment by fgerards** [comment]
any news on the subject ? Issue ist still present with TYPO3 8.7.27 & metaseo 3.0.0.
I think this is related to the url-fetching in htdocs/typo3conf/ext/metaseo/Classes/Utility/FrontendUtility.php:193: 

$tsfeAnchorPrefix = Typo3GeneralUtility::getIndpEnv('TYPO3_SITE_SCRIPT');
If a client calls the TYPO3 instance without proper useragent etc., the metaseo logic might be circumvented in a way and the GET-Params stay in the URL when added to the sitemap (TYPO3_SITE_SCRIPT might not always be set).
